docs/hlo_dumps.md

```diff
@@ -0,0 +1,219 @@
+
+1. Text dumps
+
+```shell
+XLA_FLAGS="--xla_dump_hlo_as_text --xla_dump_to=DIRECTORY_PATH"
+```
+
+2. HLO protos
+
+```shell
+XLA_FLAGS="--xla_dump_hlo_as_proto --xla_dump_to=DIRECTORY_PATH"
+```
+
+3. HLO Snapshots
+
+```shell
+XLA_FLAGS="--xla_dump_hlo_snapshots --xla_dump_to=DIRECTORY_PATH"
+```
+
+4. Graph render with graphviz server (only works well for small graphs)
+
+```shell
+XLA_FLAGS="--xla_dump_hlo_as_url --xla_dump_to=DIRECTORY_PATH"
+```
+
+5. Graph render to HTML file (only works well for small graphs)
+
+```shell
+XLA_FLAGS="--xla_dump_hlo_as_html --xla_dump_to=DIRECTORY_PATH"
+```
+
+For larger graphs, you can use `interactive_graphviz` to visualize parts of the graph.
+
+**Note:** If `--xla_dump_to` is not specified but another dumping flag is specified, it will dump to stdout. But the dump will not include binary data, e.g., proto files, to stdout.
+
+## Dump Specific Intermediate Passes
+
+In addition to the standard pre-optimized / final-optimized HLOs, you can also dump the state of HLOs after a particular compiler pass.
+
+```shell
+XLA_FLAGS="--xla_dump_hlo_pass_re=regex --xla_dump_to=DIRECTORY_PATH"
+```
+
+HLO modules will be dumped for the passes whose names match the regular expression (regex). For example, you can observe the HLOs resulting from passes related to SPMD partitioning with:
+
+```shell
+XLA_FLAGS="--xla_dump_to=DIRECTORY_PATH --xla_dump_hlo_pass_re=spmd|propagation"
+```
+
+To dump the result after every XLA pass (this will result in a lot of files), you can set:
+
+```shell
+XLA_FLAGS="--xla_dump_to=DIRECTORY_PATH --xla_dump_hlo_pass_re=.*"
+```
+
+### JAX-specific Options
+
+#### Programmatically in JAX
+
+Instead of passing flags or environment variables, you can also programmatically dump HLO using JAX’s `lower` and `compil`e APIs.
+
+Locally fetch the unoptimized original lowered HLO with:
+
+```python
+jax.jit(f).lower(*args).as_text('hlo')
+```
+
+For dumping to files during HLO compilation passes, specify:
+
+```python
+compilation_args = {
+    'xla_dump_to': DIRECTORY_PATH,
+    'xla_dump_hlo_pass_re': 'spmd|propagation', # or some other pass filter
+    ...
+    }
+
+jax.jit(f).lower(*args).compile(compilation_args)
+```
+
+#### Dump jaxprs
+
+`jaxpr`s are JAX's intermediate representation for program traces. To dump this, set the environment variables:
+
+```shell
+JAX_DUMP_IR_TO="DIRECTORY_PATH" JAX_DUMP_IR_MODES=jaxpr
+```
+
+Learn more in JAX documentation on [Exporting and serializing staged-out computations: Debugging](https://docs.jax.dev/en/latest/export/export.html#debugging).
+
+## Google Colab
+
+### Environment variables
+
+In the first executed cell of your notebook (because environment variables and command-line flags are usually only processed once, e.g., at module-import time or XLA backend initialization time), add the `XLA_FLAGS` detailed above with `os.environ`, for example:
+
+```python
+import os
+os.environ['XLA_FLAGS'] = "--xla_dump_to=CNS_DIRECTORY_PATH"
```

Comment:
Ah, we'll have to drop any reference to "CNS". That is a google internal filesystem.

Instead of `CNS_DIRECTORY_PATH`, you could say something like `YOUR_PATH`. And an example path could be simply `"/tmp"`.

You should mention that in Google Colab, you can then navigate on the left to /tmp/ in the "Files" browser. (see the screenshot I shared, from what you see, maybe you can phrase it better than me how to find the files in the file browser).